src/core/database.cpp

```diff
@@ -2594,6 +2597,94 @@ void Database::GetRefinementAngularDistributionHistogramData(long wanted_refinem
 	EndBatchSelect();
 }
 
+bool Database::UpdateSchema(std::vector<std::tuple<wxString, wxString, char>> columns) {
+	
+	CreateAllTables();
+	char format;
+	wxString column_format;
+	for (auto & column : columns) {
+		format = std::get<2>(column);
+		if (format == 't') // text
+		{
+			column_format = "TEXT";
+		}
+		else
+		if (format == 'r') // real
+		{
+			column_format = "REAL";
+		}
+		else
+		if (format == 'i') // integer
+		{
+			column_format = "INTEGER";
+		}
+		else
+		if (format == 'l') // integer
+		{
+			column_format = "INTEGER";
+		}
+		else
+		{
+			MyPrintWithDetails("Error: Unknown format character!\n");
+			column_format = "BLOB";
+		}
+		ExecuteSQL(wxString::Format("ALTER TABLE %s ADD COLUMN %s %s;",std::get<0>(column),std::get<1>(column),column_format));
+		ExecuteSQL(wxString::Format("UPDATE MASTER_SETTINGS SET CURRENT_VERSION = %i, CISTEM_VERSION_TEXT = '%s'",INTEGER_DATABASE_VERSION,CISTEM_VERSION_TEXT));
+
+	}
+	return true;
+}
+
+std::pair<std::vector<wxString>, std::vector<std::tuple<wxString, wxString, char>>> Database::CheckSchema() {
+	MyDebugAssertTrue(is_open == true, "database not open!");
+	std::vector<wxString> missing_tables;
+	std::vector<std::tuple<wxString, wxString, char>> missing_columns;
+	// Check Static Tables
+	wxArrayString return_strings;
+	int count;
+	int counter;
+	int col_counter;
+	for (auto & table : static_tables ) {
+		
+		return_strings = ReturnStringArrayFromSelectCommand(wxString::Format("SELECT name FROM sqlite_master WHERE type='table' AND name='%s';",std::get<0>(table)));
+		if (return_strings.IsEmpty()) {
+			missing_tables.push_back(std::get<0>(table));
+			continue;
+		}
+		for (col_counter = 0; col_counter < std::get<2>(table).size(); col_counter++) {
+			auto & column = std::get<2>(table)[col_counter];
+			char type = std::get<1>(table)[col_counter];
+
+			count = ReturnSingleIntFromSelectCommand(wxString::Format("SELECT COUNT(*) AS CNTREC FROM pragma_table_info('%s') WHERE name='%s';",std::get<0>(table),column));
+			if (count < 1) {
+				missing_columns.push_back(std::tuple<wxString, wxString, char>(std::get<0>(table),column,type));
+			}
+		}
+	}
+
+	for (auto & table : dynamic_tables ) {
+		
+		return_strings = ReturnStringArrayFromSelectCommand(wxString::Format("SELECT name FROM sqlite_master WHERE type='table' AND name  LIKE '%s_%';",std::get<0>(table)));
+		for (counter = 0; counter < return_strings.GetCount(); counter++) {
+			// Make sure it is not any of the static columns that happen to match
+			if (any_of(static_tables.begin(), static_tables.end(), [&](std::tuple<wxString, char *, std::vector<wxString>> & table) { return return_strings[counter].IsSameAs(std::get<0>(table)); })) {
```

Comment:
In this line you are using tabs, in the next spaces. Please set your editor to insert ONLY spaces and set it to 4 spaces/tab.

Reply:
I have it set to 4 spaces, but I think VSCode tries to match existing code so it uses tabs in existing function. 

Maybe at some point we have to make a giant commit where we run all files through a prettier..